pkg/storage/stores/shipper/indexshipper/tsdb/compactor_fd_leak_test.go

```diff
@@ -0,0 +1,242 @@
+package tsdb
+
+import (
+	"context"
+	"os"
+	"path/filepath"
+	"runtime"
+	"sync"
+	"testing"
+	"time"
+
+	"github.com/prometheus/common/model"
+	"github.com/stretchr/testify/require"
+
+	"github.com/grafana/loki/v3/pkg/storage/chunk/client/local"
+	"github.com/grafana/loki/v3/pkg/storage/chunk/client/util"
+	"github.com/grafana/loki/v3/pkg/storage/config"
+)
+
+// TestSetupBuilder_FileDescriptorLeak tests that setupBuilder properly closes
+// files immediately after use, preventing file descriptor leaks when processing
+// many source files.
+//
+// This test detects the leak by monitoring file descriptors DURING the execution
+// of setupBuilder, not after it returns. The bug causes files to accumulate during
+// the loop because defer statements keep them open until the function returns.
+func TestSetupBuilder_FileDescriptorLeak(t *testing.T) {
```

Comment:
I wonder if we actually need this test, since we are only correcting the incorrect usage of 'defer' statements within a for loop. It is beneficial to have as many tests as possible, but flaky tests are significantly more painful and annoying.

Reply:
Sure, let's only keep `TestSetupBuilder_ManyFiles` in `compactor_test.go` :ok_hand: 

That test was mainly here to highlight the issue.